crates/extensions/c8y_mapper_ext/src/inventory.rs

```diff
@@ -0,0 +1,378 @@
+    }
+
+    /// Convert a twin metadata message into Cumulocity inventory update messages.
+    /// Updating the `name` and `type` fragments are not supported.
+    /// Empty payload is mapped to a clear inventory fragment message in Cumulocity.
+    pub(crate) fn try_convert_entity_twin_data(
+        &mut self,
+        source: &EntityTopicId,
+        message: &Message,
+        fragment_key: &str,
+    ) -> Result<Vec<Message>, ConversionError> {
+        if fragment_key == "name" || fragment_key == "type" {
+            warn!("Updating the entity `name` and `type` fields via the twin/ topic channel is not supported");
+            return Ok(vec![]);
+        }
+
+        let payload = if message.payload_bytes().is_empty() {
+            JsonValue::Null
+        } else {
+            let payload = serde_json::from_slice::<JsonValue>(message.payload_bytes())?;
+            let existing = self.inventory_model.get(fragment_key);
+            if existing.is_some_and(|v| payload.eq(v)) {
+                return Ok(vec![]);
+            }
+
+            self.update_inventory_model(fragment_key.into(), payload.clone());
+            payload
+        };
+
+        let mapped_json = json!({ fragment_key: payload });
+        let mapped_message = self.inventory_update_message(source, mapped_json)?;
+        Ok(vec![mapped_message])
+    }
+
+    /// Create a Cumulocity inventory update message from a JSON fragment
+    pub(crate) fn inventory_update_message(
+        &self,
+        source: &EntityTopicId,
+        fragment_value: JsonValue,
+    ) -> Result<Message, ConversionError> {
+        let entity_external_id = self.entity_store.try_get(source)?.external_id.as_ref();
+        let inventory_update_topic = Topic::new_unchecked(&format!(
+            "{INVENTORY_MANAGED_OBJECTS_TOPIC}/{entity_external_id}"
+        ));
+
+        Ok(Message::new(
+            &inventory_update_topic,
+            fragment_value.to_string(),
+        ))
+    }
+
+    /// Create the inventory update message to update the `type` of the main device
+    pub(crate) fn inventory_device_type_update_message(&self) -> Result<Message, ConversionError> {
+        let device_data = C8yDeviceDataFragment::from_type(&self.device_type)?;
+        let device_type_fragment = device_data.to_json()?;
+
+        self.inventory_update_message(&self.device_topic_id, device_type_fragment)
+    }
+
+    /// Return the contents of inventory.json file as a `JsonValue`
+    pub(crate) fn get_inventory_fragments(
+        inventory_file_path: &Path,
+    ) -> Result<JsonValue, ConversionError> {
+        let agent_fragment = C8yAgentFragment::new()?;
+        let json_fragment = agent_fragment.to_json()?;
+
+        match Self::read_json_from_file(inventory_file_path) {
+            Ok(mut json) => {
+                json.as_object_mut()
+                    .ok_or(ConversionError::FromOptionError)?
+                    .insert(
+                        "c8y_Agent".to_string(),
```

Comment:
Why are `c8y_Agent` and `type` treated differently?

- The former is inserted into `self.inventory_model`, the latter is not.
- The former is published over a `te` twin topic, the latter is not.

Reply:
> * The former is inserted into `self.inventory_model`, the latter is not

The `name` and `type` are inserted into the inventory in the `CumulocityConverter` constructor. It was just accidentally getting lost with this buggy line: `self.inventory_model = inventory_base.clone();` which I'll revert.

> The former is published over a te twin topic, the latter is not.

`type` and `name` are not published for the same reason why updating them via `twin/` topics are not supported. They must only be present in the entity registration message and we don't want another copy of it at `twin/name` and `twin/type` topics.

Reply:
Aren't `name` and `c8y_Agent.name` the same thing? I believe my confusion is coming from that.

Reply:
No, `name` here is the device/service name and `c8y_Agent.name` is the name of the agent. When thin-edge connects to C8y, the `c8y_Agent.name` is `thin-edge.io`.